### PR TITLE
M1

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
+ */
+var config = require('bedrock').config;
+var path = require('path');
+require('bedrock-agreement');
+require('bedrock-identity');
+require('bedrock-key');
+require('bedrock-validation');
+
```
Comment (mattcollier) [inline lib/config.js]: I think these requires need to be in index.js.

**Comment by mattcollier** [comment]
@dlongley ready for a look see.

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+var BedrockError = bedrock.util.BedrockError;
+var validate = require('bedrock-validation').validate;
+var logger = bedrock.loggers.get('app');
+
+// pull in the default config
+require('./config');
+
+// add routes
+bedrock.events.on('bedrock-express.configure.routes', addRoutes);
```
Comment (dlongley) [inline lib/index.js]: Add `require('bedrock-express');` above.

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+function addRoutes(app) {
```
Comment (dlongley) [inline lib/index.js]: Can probably just move this up and do: `app => {`

File: lib/index.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  // warn if this module is included but all functionality is disabled
+  if(!config['website-user-http'].features.join) {
+    logger.warning('user joining is disabled, enable by setting ' +
+      'bedrock.config.website-user-http.enableJoining to true');
```
Comment (dlongley) [inline lib/index.js]: Fix warning; has the wrong config var to change.

File: lib/index.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  // creating new identities must be enabled via flags
+  if(config['website-user-http'].features.join) {
+    app.post(config['website-user-http'].routes.join,
+      validate('services.join.postJoin'),
+      function(req, res, next) {
+        async.waterfall([
```
Comment (dlongley) [inline lib/index.js]: All of this could be ES6'd up. Also, use `async.auto` not `async.waterfall` for clarity.

File: lib/index.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            });
+          }
+        ], function(err) {
+          if(err) {
+            return next(err);
+          }
+        });
+      });
+  }
+
```
Comment (dlongley) [inline lib/index.js]: Remove extra space.

File: lib/index.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+/**
+ * Identity creation service. Used by normal and testing services.
+ */
+function _createIdentity(req, callback) {
+  var identityId = req.body.id;
```
Comment (dlongley) [inline lib/index.js]: `const` ... etc, I'm going to stop commenting on ES6 stuff -- just needs to be addressed in general.

File: lib/index.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    // result details
+    var details = {identity: results.createIdentity};
+    callback(null, details);
+  });
+}
+
+// set reasonable defaults for newly created identities
+bedrock.events.on('bedrock-identity.insert', function(options, callback) {
+
```
Comment (dlongley) [inline lib/index.js]: Remove empty line.

**Review by dlongley — CHANGES_REQUESTED** [review]
AFAICT this looks ok functionality wise, just need a few tweaks to bring it up to speed on style.

File: lib/config.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Bedrock Key HTTP Module Configuration
+ *
+ * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
+ */
+var config = require('bedrock').config;
+var path = require('path');
```
Comment (dlongley) [inline lib/config.js]: Should be `const`

File: schemas/services.join.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        }
+      }
+    }
+  },
+  additionalProperties: false
+};
+
+module.exports.postJoin = function() {
+  return postJoin;
+};
```
Comment (davidlehn) [inline schemas/services.join.js]: Could one-line this with ES2015:
```js
module.exports.postJoin = () => postJoin;
```

**Comment by mattcollier** [comment]
@dlongley @davidlehn ready for another pass.

File: schemas/services.join.js (hunk 10 above)
Comment (dlongley) [reply schemas/services.join.js]: Why not just `module.exports.postJoin = postJoin;`?

Comment (dlongley) [reply schemas/services.join.js]: The reason for it to be a function is to allow schema changes, but we don't even permit that here.

**Review by dlongley — APPROVED** [review]

**Review by davidlehn — APPROVED** [review]

**Pull request merged**